app/src/main/java/org/stepik/android/presentation/course_list/mapper/CourseListStateMapper.kt

```diff
@@ -95,7 +97,48 @@ constructor() {
         )
     }
 
-    fun mapEnrolledCourseListItemState(state: CourseListView.State, courseListItemEnrolled: CourseListItem.Data): CourseListView.State =
+    fun mapUserCourseRemoveState(oldState: CourseListUserView.State.Data, oldCourseListState: CourseListView.State.Content, courseId: Long): CourseListUserView.State.Data {
+        val userCoursesUpdate = oldState.userCourses.mapNotNull {
+            if (it.course == courseId) {
+                null
+            } else {
+                it
+            }
+        }
+        val index = oldCourseListState.courseListDataItems
+            .indexOfFirst { it.course.id == courseId }
+
+        val newItems = oldCourseListState.courseListDataItems.mutate { removeAt(index) }
+
+        val courseListViewState = if (newItems.isNotEmpty()) {
+            oldCourseListState.copy(
+                courseListDataItems = newItems,
+                courseListItems = newItems
+            )
+        } else {
+            CourseListView.State.Empty
+        }
+
+        return oldState.copy(
+            userCourses = userCoursesUpdate,
+            courseListViewState = courseListViewState
+        )
+    }
+
+    fun mapUserCourseAddState(
+        oldState: CourseListUserView.State.Data,
+        userCourse: UserCourse,
+        courseListItem: CourseListItem.Data
+    ): CourseListUserView.State.Data {
+        val userCoursesUpdated = listOf(userCourse) + oldState.userCourses
```

Comment:
1. courses should be correctly sorted
2. will course item duplicate in case of next page?